File: backend/app/api/v1/endpoints/automation.py
"""
自动化命令API路由
"""
from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List
import logging

from app.schemas.automation import (
    AutomationCommandRequest,
    AutomationCommandResponse,
    CommandConfirmationRequest,
    CommandListResponse,
    CommandTemplate,
    CommandStatus,
    CommandType
)
from app.services.automation_service import automation_service
from app.core.response import APIResponse

logger = logging.getLogger(__name__)

router = APIRouter(tags=["自动化命令"])


@router.post("/commands", response_model=dict)
async def create_command(request: AutomationCommandRequest):
    """创建自动化命令"""
    try:
        command = await automation_service.create_command(request)
        logger.info(f"API: 创建命令成功 - {command.command_id}")
        return APIResponse.success(data=command.model_dump())
    except Exception as e:
        logger.error(f"API: 创建命令失败 - {e}")
        return APIResponse.error(code=500, msg=f"创建命令失败: {e}")


@router.post("/commands/{command_id}/confirm", response_model=dict)
async def confirm_command(command_id: str, confirmation: CommandConfirmationRequest):
    """确认命令执行"""
    try:
        if confirmation.command_id != command_id:
            raise ValueError("命令ID不匹配")
        
        command = await automation_service.confirm_command(confirmation)
        logger.info(f"API: 命令确认成功 - {command_id}, 确认: {confirmation.confirmed}")
        return APIResponse.success(data=command.model_dump())
    except ValueError as e:
        logger.warning(f"API: 命令确认失败 - {e}")
        return APIResponse.error(code=400, msg=str(e))
    except Exception as e:
        logger.error(f"API: 命令确认异常 - {e}")
        return APIResponse.error(code=500, msg=f"确认命令失败: {e}")


@router.get("/commands/{command_id}", response_model=dict)
async def get_command(command_id: str):
    """获取单个命令信息"""
    try:
        command = automation_service.get_command(command_id)
        if not command:
            return APIResponse.error(code=404, msg="命令不存在")
        
        return APIResponse.success(data=command.model_dump())
    except Exception as e:
        logger.error(f"API: 获取命令失败 - {e}")
        return APIResponse.error(code=500, msg=f"获取命令失败: {e}")


@router.get("/commands", response_model=dict)
async def get_commands(
    status: Optional[CommandStatus] = Query(None, description="命令状态"),
    command_type: Optional[CommandType] = Query(None, description="命令类型"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(10, ge=1, le=100, description="每页数量")
):
    """获取命令列表"""
    try:
        result = automation_service.get_commands(
            status=status,
            command_type=command_type,
            page=page,
            page_size=page_size
        )
        
        # 转换为可序列化的格式
        commands_data = []
        for cmd in result["commands"]:
            commands_data.append(cmd.model_dump())
        
        response_data = {
            "commands": commands_data,
            "total": result["total"],
            "page": result["page"],
            "page_size": result["page_size"]
        }
        
        return APIResponse.success(data=response_data)
    except Exception as e:
        logger.error(f"API: 获取命令列表失败 - {e}")
        return APIResponse.error(code=500, msg=f"获取命令列表失败: {e}")


@router.delete("/commands/{command_id}", response_model=dict)
async def cancel_command(command_id: str):
    """取消命令执行"""
    try:
        command = await automation_service.cancel_command(command_id)
        logger.info(f"API: 取消命令成功 - {command_id}")
        return APIResponse.success(data=command.model_dump())
    except ValueError as e:
        logger.warning(f"API: 取消命令失败 - {e}")
        return APIResponse.error(code=400, msg=str(e))
    except Exception as e:
        logger.error(f"API: 取消命令异常 - {e}")
        return APIResponse.error(code=500, msg=f"取消命令失败: {e}")


@router.get("/templates", response_model=dict)
async def get_templates():
    """获取命令模板列表"""
    try:
        templates = automation_service.get_templates()
        templates_data = [template.model_dump() for template in templates]
        return APIResponse.success(data=templates_data)
    except Exception as e:
        logger.error(f"API: 获取模板列表失败 - {e}")
        return APIResponse.error(code=500, msg=f"获取模板列表失败: {e}")


@router.get("/templates/{template_id}", response_model=dict)
async def get_template(template_id: str):
    """获取单个命令模板"""
    try:
        template = automation_service.get_template(template_id)
        if not template:
            return APIResponse.error(code=404, msg="模板不存在")
        
        return success_response(data=template.model_dump())
    except Exception as e:
        logger.error(f"API: 获取模板失败 - {e}")
        return APIResponse.error(code=500, msg=f"获取模板失败: {e}")


@router.post("/templates/{template_id}/execute", response_model=dict)
async def execute_template_command(
    template_id: str,
    parameters: dict,
    operator_id: Optional[str] = Query(None, description="操作员ID"),
    workstation_id: Optional[str] = Query(None, description="工位ID")
):
    """根据模板执行命令"""
    try:
        command = await automation_service.execute_template_command(
            template_id=template_id,
            parameters=parameters,
            operator_id=operator_id,
            workstation_id=workstation_id
        )
        
        logger.info(f"API: 模板命令创建成功 - 模板: {template_id}, 命令: {command.command_id}")
        return APIResponse.success(data=command.model_dump())
    except ValueError as e:
        logger.warning(f"API: 模板命令创建失败 - {e}")
        return APIResponse.error(code=400, msg=str(e))
    except Exception as e:
        logger.error(f"API: 模板命令创建异常 - {e}")
        return APIResponse.error(code=500, msg=f"执行模板命令失败: {e}")

File: frontend/src/stores/automation.ts
/**
 * 自动化命令状态管理
 */
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { ElMessage, ElMessageBox } from 'element-plus'
import AutomationAPI, { 
  type AutomationCommand, 
  type CommandTemplate, 
  type AutomationCommandRequest,
  type CommandListParams 
} from '@/api/automation'

export const useAutomationStore = defineStore('automation', () => {
  // 状态
  const commands = ref<AutomationCommand[]>([])
  const templates = ref<CommandTemplate[]>([])
  const currentCommand = ref<AutomationCommand | null>(null)
  const loading = ref(false)
  const executing = ref(false)

  // 计算属性
  const pendingCommands = computed(() => 
    commands.value.filter(cmd => cmd.status === 'pending')
  )

  const executingCommands = computed(() => 
    commands.value.filter(cmd => cmd.status === 'executing')
  )

  const recentCommands = computed(() => 
    commands.value
      .sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime())
      .slice(0, 10)
  )

  /**
   * 加载命令模板
   */
  const loadTemplates = async () => {
    try {
      loading.value = true
      const data = await AutomationAPI.getTemplates()
      templates.value = data || []
    } catch (error) {
      console.error('加载模板失败:', error)
      ElMessage.error('加载模板失败')
    } finally {
      loading.value = false
    }
  }

  /**
   * 加载命令列表
   */
  const loadCommands = async (params?: CommandListParams) => {
    try {
      loading.value = true
      const data = await AutomationAPI.getCommands(params)
      commands.value = data?.commands || []
    } catch (error) {
      console.error('加载命令列表失败:', error)
      ElMessage.error('加载命令列表失败')
    } finally {
      loading.value = false
    }
  }

  /**
   * 创建自动化命令
   */
  const createCommand = async (request: AutomationCommandRequest): Promise<AutomationCommand | null> => {
    try {
      executing.value = true
      const command = await AutomationAPI.createCommand(request)
      commands.value.unshift(command)
      currentCommand.value = command
      
      ElMessage.success('命令创建成功')
      
      // 如果需要确认，显示确认弹窗
      if (request.requires_confirmation && command.status === 'pending') {
        await showConfirmationDialog(command)
      }
      
      return command
    } catch (error) {
      console.error('创建命令失败:', error)
      ElMessage.error('创建命令失败')
      return null
    } finally {
      executing.value = false
    }
  }

  /**
   * 根据模板执行命令
   */
  const executeTemplate = async (
    templateId: string, 
    parameters: Record<string, any>,
    operatorId?: string,
    workstationId?: string
  ): Promise<AutomationCommand | null> => {
    try {
      executing.value = true
      const command = await AutomationAPI.executeTemplateCommand(
        templateId, 
        parameters, 
        operatorId, 
        workstationId
      )
      
      commands.value.unshift(command)
      currentCommand.value = command
      
      ElMessage.success('命令创建成功')
      
      // 如果需要确认，显示确认弹窗
      const template = templates.value.find(t => t.template_id === templateId)
      if (template?.requires_confirmation && command.status === 'pending') {
        await showConfirmationDialog(command)
      }
      
      return command
    } catch (error) {
      console.error('执行模板命令失败:', error)
      ElMessage.error('执行模板命令失败')
      return null
    } finally {
      executing.value = false
    }
  }

  /**
   * 显示确认弹窗
   */
  const showConfirmationDialog = async (command: AutomationCommand): Promise<boolean> => {
    try {
      await ElMessageBox.confirm(
        `确定要执行命令吗？`,
        '命令确认',
        {
          confirmButtonText: '确认执行',
          cancelButtonText: '取消',
          type: 'warning',
          customClass: 'automation-confirm-dialog',
          beforeClose: (action, instance, done) => {
            if (action === 'confirm') {
              confirmCommand(command.command_id, true)
            } else {
              confirmCommand(command.command_id, false)
            }
            done()
          }
        }
      )
      return true
    } catch (error) {
      // 用户取消
      await confirmCommand(command.command_id, false)
      return false
    }
  }

  /**
   * 确认命令执行
   */
  const confirmCommand = async (commandId: string, confirmed: boolean, notes?: string) => {
    try {
      const command = await AutomationAPI.confirmCommand(commandId, {
        command_id: commandId,
        confirmed,
        operator_notes: notes
      })
      
      // 更新本地命令状态
      const index = commands.value.findIndex(cmd => cmd.command_id === commandId)
      if (index !== -1) {
        commands.value[index] = command
      }
      
      if (currentCommand.value?.command_id === commandId) {
        currentCommand.value = command
      }
      
      if (confirmed) {
        ElMessage.success('命令已确认执行')
        // 开始轮询检查执行状态
        pollCommandStatus(commandId)
      } else {
        ElMessage.info('命令已取消')
      }
    } catch (error) {
      console.error('确认命令失败:', error)
      ElMessage.error('确认命令失败')
    }
  }

  /**
   * 轮询命令执行状态
   */
  const pollCommandStatus = async (commandId: string) => {
    const pollInterval = setInterval(async () => {
      try {
        const command = await AutomationAPI.getCommand(commandId)
        
        // 更新本地状态
        const index = commands.value.findIndex(cmd => cmd.command_id === commandId)
        if (index !== -1) {
          commands.value[index] = command
        }
        
        if (currentCommand.value?.command_id === commandId) {
          currentCommand.value = command
        }
        
        // 如果命令完成，停止轮询
        if (['success', 'failed', 'cancelled'].includes(command.status)) {
          clearInterval(pollInterval)
          
          if (command.status === 'success') {
            ElMessage.success('命令执行成功')
          } else if (command.status === 'failed') {
            ElMessage.error(`命令执行失败: ${command.error_message}`)
          }
        }
      } catch (error) {
        console.error('轮询命令状态失败:', error)
        clearInterval(pollInterval)
      }
    }, 2000) // 每2秒轮询一次
    
    // 30秒后停止轮询
    setTimeout(() => {
      clearInterval(pollInterval)
    }, 30000)
  }

  /**
   * 取消命令
   */
  const cancelCommand = async (commandId: string) => {
    try {
      const command = await AutomationAPI.cancelCommand(commandId)
      
      // 更新本地状态
      const index = commands.value.findIndex(cmd => cmd.command_id === commandId)
      if (index !== -1) {
        commands.value[index] = command
      }
      
      if (currentCommand.value?.command_id === commandId) {
        currentCommand.value = command
      }
      
      ElMessage.success('命令已取消')
    } catch (error) {
      console.error('取消命令失败:', error)
      ElMessage.error('取消命令失败')
    }
  }

  /**
   * 刷新命令状态
   */
  const refreshCommand = async (commandId: string) => {
    try {
      const command = await AutomationAPI.getCommand(commandId)
      
      // 更新本地状态
      const index = commands.value.findIndex(cmd => cmd.command_id === commandId)
      if (index !== -1) {
        commands.value[index] = command
      }
      
      if (currentCommand.value?.command_id === commandId) {
        currentCommand.value = command
      }
    } catch (error) {
      console.error('刷新命令状态失败:', error)
    }
  }

  /**
   * 获取状态显示文本
   */
  const getStatusText = (status: string): string => {
    const statusMap: Record<string, string> = {
      pending: '等待确认',
      confirmed: '已确认',
      executing: '执行中',
      success: '执行成功',
      failed: '执行失败',
      cancelled: '已取消'
    }
    return statusMap[status] || status
  }

  /**
   * 获取状态颜色类型
   */
  const getStatusType = (status: string): string => {
    const typeMap: Record<string, string> = {
      pending: 'warning',
      confirmed: 'info',
      executing: 'primary',
      success: 'success',
      failed: 'danger',
      cancelled: 'info'
    }
    return typeMap[status] || 'info'
  }

  /**
   * 获取命令类型显示文本
   */
  const getCommandTypeText = (type: string): string => {
    const typeMap: Record<string, string> = {
      system: '系统命令',
      device: '设备命令',
      test: '测试命令',
      maintenance: '维护命令'
    }
    return typeMap[type] || type
  }

  return {
    // 状态
    commands,
    templates,
    currentCommand,
    loading,
    executing,
    
    // 计算属性
    pendingCommands,
    executingCommands,
    recentCommands,
    
    // 方法
    loadTemplates,
    loadCommands,
    createCommand,
    executeTemplate,
    confirmCommand,
    cancelCommand,
    refreshCommand,
    
    // 工具方法
    getStatusText,
    getStatusType,
    getCommandTypeText
  }
})